### ISSUE TITLE
FK with Different DataType than PK Ask Question

### ISSUE THREAD

**Issue by jaredcnance** [dotnet/efcore#11112]
I'm working on an old database that is being used by lots of services and the original creators defined the PK of one table as `smallint` and a FK to that table as an `int`. Is it at all possible for me to handle this case? I recognize that this is a consequence of poor design, but I cannot feasibly alter the database. Other applications depend on the current datatypes and would break if I altered the database. I would still like to use Navigation properties if possible.

### Steps to reproduce

```c#
public class Database {
  public short DatabaseId { get; set; } // PK
  public Database { get; set; }
}

public class Client {
  public int ClientId { get; set; }
  public int DatabaseId { get; set; } // FK
  public Database Database { get; set; }
}
```

### A few attempts:

```csharp
public int DatabaseId { get; set; }
[ForeignKey("DatabaseId")] public Database Database { get; set; }
```

throws:

```
System.InvalidOperationException: The relationship from 'Client.Database' to 'Database.Client' with foreign key properties {'DatabaseId' : int} cannot target the primary key {'DatabaseId' : short} because it is not compatible. Configure a principal key or a set of compatible foreign key properties for this relationship.
```

### Fluent API 

```
modelBuilder.Entity<Client>(table => {    
    table.HasOne(p => p.Database)
        .WithOne(i => i.Client)
        .HasForeignKey<Client>(c => c.DatabaseId)
        .HasPrincipalKey<Database>(d => d.DatabaseID);
});
```

throws

```
System.InvalidOperationException: The types of the properties specified for the foreign key {'DatabaseId'} on entity type 'Client' do not match the types of the properties in the principal key {'DatabaseID'} on entity type 'Database'.
```

### Using Shadow Properties

```c#
modelBuilder.Entity<Client>(table => {
    table.Property<int>("DatabaseId");

    table.HasOne(p => p.Database)
        .WithOne(i => i.Client)
        .HasForeignKey("DatabaseId")
        .HasPrincipalKey<Database>(d => d.DatabaseId);
});
```

throws:

```
System.InvalidOperationException: You are configuring a relationship between 'Client' and 'Database' but have specified a foreign key on 'DatabaseId'. The foreign key must be defined on a type that is part of the relationship.
```

### Specifying the column type

```c#
[Column(TypeName = "int")]
public short DatabaseId { get; set; }
```

```
 System.InvalidOperationException: An exception occurred while reading a database value for property 'Client.DatabaseId'. The expected type was 'System.Int16' but the actual value was of type 'System.Int32'. ---> System.InvalidCastException: Unable to cast object of type 'System.Int32' to type 'System.Int16'.
```

I also tried going the other way and defining it as an `int`.

I'm looking for a workaround, but if it is simply not possible, please let me know. Thanks!

### Further technical details
EF Core version: 2.0.1
Database Provider: Microsoft.EntityFrameworkCore.SqlServer 2.0.1
Operating system: OSX
IDE: Visual Studio Code

**Comment by ajcvickers** [comment]
@jaredcnance At least with SQL Server, you should be able to just lie in your model and make the property an int:
```C#
public class Database {
  public int DatabaseId { get; set; } // PK
  public Database { get; set; }
}
```
I tried this and it worked for me. Beyond that, you should be able to use value converters in 2.1, although there may be some issues in the preview build--I filed #11117 to track this.

**Comment by ajcvickers** [comment]
Triage: closing this as it should be possible to do with ValueConverters, but #11117 (which we plan to fix for 2.1) is blocking it right now.
